### ISSUE TITLE
Make ASN.1 integer writing independent of MPI type

### ISSUE THREAD

**Issue by gilles-peskine-arm** [Mbed-TLS/mbedtls#9372]
In Mbed TLS 4.0, `mbedtls_mpi` is becoming private (to be removed in 4.0 or a later 4.x in favor of the representations in `bignum_core.h` and `bignum_mod.h`). We need to change the interface of the ASN.1 integer writing function accordingly.

The typical usage will be to encode an integer exported with [`mbedtls_mpi_core_write_be`](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.6.0/library/bignum_core.h#L257). What should the interface be? Note that ASN.1 encoding needs to know the exact bit-length, not just the exact byte-length, since the top bit of the first byte of the content must be 0 (it's a sign bit), e.g. 0x007f0000 is encoded as `02037f0000` but 0x00800000 as `020400080000`. That makes it difficult to combine an ASN.1 writing function with `mbedtls_mpi_core_write_be`.

Definition of done:

* The new function is implemented and tested.
* We have at least one test that uses `mbedtls_asn1_write_integer` with a bignum.
* Move `mbedtls_asn1_write_mpi` to an internal header (`bignum.h`, which will be made non-public?), as it is still useful for modules that use `mbedtls_mpi` internally.

**Comment by gilles-peskine-arm** [comment]
### Interface proposal: reserve and copy

```
unsigned char *octets_head;
size_t octets_len;
size_t bits = mbedtls_mpi_core_bitlen(A, A_limbs); // or mbedtls_mpi_bitlen(A)
mbedtls_asn1_write_mpi(p, start, bits, &octets_head, &octets_len);
mbedtls_mpi_core_write_be(A, A_limbs, octets_head, octets_len);
```

Downside: two functions, requires the caller to pass matching arguments.

### Interface proposal: shared knowledge

```
mbedtls_asn1_write_mpi_core(p, start, A, A_limbs);
```

Downside: the same code needs to understand both the ASN.1 encoding and the bignum representation.

**Comment by davidhorstmann-arm** [comment]
> Note that ASN.1 encoding needs to know the exact bit-length, not just the exact byte-length, since the top bit of the first byte of the content must be 0

I think that ASN.1 also needs the bit length as overlong-encoding of integers is not allowed (i.e. we always have to have the minimum possible number of leading zeroes). This means that we'll always have to inspect the integer's value irrespective of the number of limbs allocated for it.

Incidentally I notice that this makes it near-impossible to implement constant-time parsing of keys.

**Comment by mpg** [comment]
> Incidentally I notice that this makes it near-impossible to implement constant-time parsing of keys.

Yes, people who specify variable-length encoding for crypto material must be either blissfully unaware of side channels, or very deeply hate crypto implementers. Joke aside, historically I guess it's the first one, I mean remote timing attacks have only [been (known to be) practical since 2003](https://dl.acm.org/doi/10.5555/1251353.1251354) and several crypto standards used today are older, for example PKCS#1 [started in the early 90s](https://en.wikipedia.org/wiki/PKCS_1#Version_history) (and might have older roots).

Case in point, TLS was using variable-length encoding for FFDH until 1.2, leading to the [Raccoon attack](https://raccoon-attack.com/) which calls it "a TLS specification side channel" (IMO meaning the fault lies with the spec, not the implementations), and this was changed to fixed-length encoding in 1.3 (ECDH always used a fixed-length encoding). Again, according to their FAQ this is a relic form SSL3, which again [is from the 90s](https://en.wikipedia.org/wiki/Transport_Layer_Security#History_and_development) - crypto standards from that decade did not age as well as its music ;)
